cmd/ocsp-updater/main.go

```diff
@@ -0,0 +1,202 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package main
+
+import (
+	"crypto/x509"
+	"database/sql"
+	"fmt"
+	"math"
+	"time"
+
+	// Load both drivers to allow configuring either
+	_ "github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/go-sql-driver/mysql"
+	_ "github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/mattn/go-sqlite3"
+
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/cactus/go-statsd-client/statsd"
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/codegangsta/cli"
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/streadway/amqp"
+	gorp "github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp.v1"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/core"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/rpc"
+	"github.com/letsencrypt/boulder/sa"
+)
+
+const ocspResponseLimit int = 128
+
+func setupClients(c cmd.Config) (rpc.CertificateAuthorityClient, chan *amqp.Error) {
+	ch := cmd.AmqpChannel(c.AMQP.Server)
+	closeChan := ch.NotifyClose(make(chan *amqp.Error, 1))
+
+	cac, err := rpc.NewCertificateAuthorityClient("OCSP->CA", c.AMQP.CA.Server, ch)
+	cmd.FailOnError(err, "Unable to create CA client")
+
+	return cac, closeChan
+}
+
+func processResponse(cac rpc.CertificateAuthorityClient, tx *gorp.Transaction, serial string) error {
+	certObj, err := tx.Get(core.Certificate{}, serial)
+	if err != nil {
+		return err
+	}
+	statusObj, err := tx.Get(core.CertificateStatus{}, serial)
+	if err != nil {
+		return err
+	}
+
+	cert := certObj.(*core.Certificate)
+	status := statusObj.(*core.CertificateStatus)
+
+	_, err = x509.ParseCertificate(cert.DER)
+	if err != nil {
+		return err
+	}
+
+	signRequest := core.OCSPSigningRequest{
+		CertDER:   cert.DER,
+		Reason:    status.RevokedReason,
+		Status:    string(status.Status),
+		RevokedAt: status.RevokedDate,
+	}
+
+	ocspResponse, err := cac.GenerateOCSP(signRequest)
+	if err != nil {
+		return err
+	}
+
+	timeStamp := time.Now()
+
+	// Record the response.
+	ocspResp := &core.OcspResponse{Serial: serial, CreatedAt: timeStamp, Response: ocspResponse}
+	err = tx.Insert(ocspResp)
+	if err != nil {
+		return err
+	}
+
+	// Reset the update clock
+	status.OCSPLastUpdated = timeStamp
+	_, err = tx.Update(status)
+	if err != nil {
+		return err
+	}
+
+	// Done
+	return nil
+}
+
+func findStaleResponses(cac rpc.CertificateAuthorityClient, dbMap *gorp.DbMap, oldestLastUpdatedTime time.Time, responseLimit int) error {
+	log := blog.GetAuditLogger()
+
+	// If there are fewer than this many days left before the currently-signed
+	// OCSP response expires, sign a new OCSP response.
+	var certificateStatus []core.CertificateStatus
+	_, err := dbMap.Select(&certificateStatus,
+		`SELECT cs.* FROM certificateStatus AS cs
+		 WHERE cs.ocspLastUpdated < ?
+		 ORDER BY cs.ocspLastUpdated ASC
+		 LIMIT ?`, oldestLastUpdatedTime, responseLimit)
```

Comment:
You're going to make way more OCSP responses than you need to with this query.  You should only be generating OCSP information for certificates that are not expired.  I think we need to add a "notAfter" column to the certificateStatus table, so that we can select on that here.  I'm not sure how to add a commit to this PR, but here's a micro-patch in an ad-hoc format:

```
core/objects.go: type CertificateStatus struct {
+	// notAfter: The notAfter time of the certificate.  This is simply copied
+	// from the certificate itself, but simplifies queries needed to make CSRs
+	// and OCSP responses
+	NotAfter time.Time `db:"notAfter"`

sa/storage-authority.go: AddCertificate(
 		Serial:             serial,
+		NotAfter:           parsedCertificate.NotAfter,
 		RevokedDate:        time.Time{},
```

If you make that change, then it seems like you should be able to just add a clause to that SELECT statement of the form:

```
cs.notAfter > ?
```

... where the `?` is populated with something like `time.Now()`.